dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/AsyncToolCancellationTest.java

```diff
@@ -0,0 +1,124 @@
+        WebArchive war = ShrinkWrap.create(WebArchive.class, "asyncToolCancellationTest.war")
+                                   .addPackage(AsyncTools.class.getPackage())
+                                   .addPackage(AwaitToolServlet.class.getPackage());
+
+        ShrinkHelper.exportDropinAppToServer(server, war, SERVER_ONLY);
+
+        server.startServer();
+        executor = Executors.newSingleThreadExecutor();
+        assertNotNull(server.waitForStringInLog("MCP server endpoint: .*/mcp$"));
+    }
+
+    @AfterClass
+    public static void teardown() throws Exception {
+        executor.shutdown();
+        server.stopServer(EXPECTED_ERROR);
+    }
+
+    @Test
+    public void testCancellationToolAsync() throws Exception {
+
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        Callable<String> threadCallingTool = () -> {
+            try {
+                String request = """
+                                  {
+                                  "jsonrpc": "2.0",
+                                  "id": "2",
+                                  "method": "tools/call",
+                                  "params": {
+                                    "name": "asyncCancellationTool",
+                                    "arguments": {
+                                      "latchName": "testCancellationToolAsync"
+                                    }
+                                  }
+                                }
+                                """;
+
+                latch.countDown();
+                return client.callMCP(request);
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        };
+
+        Future<String> future = executor.submit(threadCallingTool);
+
+        String cancellationRequestNotification = """
+                          {
+                          "jsonrpc": "2.0",
+                          "method": "notifications/cancelled",
+                          "params": {
+                            "requestId": "2",
+                            "reason": "no longer needed"
+                          }
+                        }
+                        """;
+        latch.await();
+        // Call AwaitToolServlet to wait for the tool to start running. Adds path param "strId" to specify which countdown latch to use
+        new HttpRequest(server, "/asyncToolCancellationTest/awaitTool/testCancellationToolAsync").run(String.class);
+
+        client.callMCPNotification(server, "/asyncToolCancellationTest", cancellationRequestNotification);
+
+        String response = future.get(10, TimeUnit.SECONDS);
+
+        String expectedResponseString = """
+                        {"id":"2","jsonrpc":"2.0","result":{"content":[{"text":"CWMCM0011E: An internal server error occurred while running the tool.", "type":"text"}],"isError":true}}
+                        """;
+        JSONAssert.assertEquals(expectedResponseString, response, true);
```

Comment:
Why do we get an internal server error response when we cancel a tool?

That doesn't seem right, I'll raise an issue for this.